infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -17,49 +17,117 @@
+    fputc('\n', stderr);          \
   } while (0)
 #else
 #define debug_log(...)
 #endif
 
-#define fatal_log(...) \
-  do { \
+#define fatal_log(...)            \
+  do {                            \
     fprintf(stderr, __VA_ARGS__); \
-    fputc('\n', stderr); \
-    exit(EXIT_FAILURE); \
+    fputc('\n', stderr);          \
+    exit(EXIT_FAILURE);           \
   } while (0)
 
 // The magic string that we'll use to detect full control over the command
 // executed.
 const std::string kTripWire = "/tmp/tripwire";
+// Shell injection bug confirmed with /tmp/tripwire.
+const std::string kInjectionError = "Shell injection";
+// Shell corruption bug detected based on syntax error.
+const std::string kCorruptionError = "Shell corruption";
+
 // The PID of the root process we're fuzzing.
 pid_t g_root_pid;
+// Assuming the longest pathname is "/bin/bash".
+constexpr int kShellPathnameLength = 10;
+// Assuming the syntax error pattern is
+// within the first 100 chars of the write buffer.
+constexpr int kErrorMessageLength = 100;
+
+// Syntax error messages of each shell.
+const std::map<std::string, std::set<std::string>> kShellSyntaxErrors = {
+    {"sh",
+     {
+         ": command not found",  // General
+         ": syntax error",       // Unfinished " or ' or ` or if, leading | or ;
+         ": missing"             // Unfinished [
+         ": event not found",    // ! leads large numbers
+         ": No such file",       // Leading < or /
+     }},
+    {"bash",
+     {
+         ": command not found",  // General
+         ": syntax error",       // Unfinished " or ' or ` or if, leading | or ;
+         ": missing"             // Unfinished [
+         ": event not found",    // ! leads large numbers
+         ": No such file",       // Leading < or /
+     }},
+    {"csh",
+     {
+         ": Command not found",       // General
+         "Unmatched",                 // Unfinished " or ' or `, leading ;
+         ": Missing",                 // Unfinished {
+         "Too many",                  // Unfinished (
+         "No match",                  // Leading [
+         "Invalid null command",      // Leading | or < or >
+         "Missing name for redirect"  // Single < or >
+         ": No match",                // Leading ? or [ or *
+         "Modifier failed",           // Leading ^
+         ": No such job",             // Leading %
+         ": Undefined variable",      // Containing $
+         ": Event not found",         // ! leads large numbers
+     }},
+    {"dash",
+     {
+         ": not found",     // General
+         ": Syntax error",  // Unfinished " or ' or ` or if, leading | or ; or &
+         ": missing"        // Unfinished [
+         ": No such file",  // Leading <
+     }},
+    {"zsh",
+     {
+         ": command not found",                // General
+         ": syntax error",                     // Unfinished " or ' or `
+         " expected"                           // Unfinished [
```

Comment:
this seems a bit too generic. Is there anything else we can match on?

Reply:
Ah, yes!
Given this is a special case for `[`, I will expand this pattern to `: missing ]`, or use the full output:
```shell
dash: 1: [: missing ]
```